server/svix-server/src/core/idempotency.rs

```diff
@@ -0,0 +1,426 @@
+// SPDX-FileCopyrightText: © 2022 Svix Authors
+// SPDX-License-Identifier: MIT
+
+//! Defines idempotency middleware for the Axum server which first looks up the given key for an
+//! existing response before routing to the given endpoint's function, and caches any such results
+//! such that subsequent requests to that endpoint with the same key will return the same response.
+//!
+//! Responses are cached for twelve hours by default.
+
+use std::{collections::HashMap, convert::Infallible, future::Future, pin::Pin, time::Duration};
+
+use axum::{
+    body::{Body, BoxBody, HttpBody},
+    http::{Request, Response, StatusCode},
+    response::IntoResponse,
+    Json,
+};
+use blake2::{Blake2b512, Digest};
+use serde::{Deserialize, Serialize};
+use tower::Service;
+
+use super::cache::{CacheKey, CacheValue, RedisCache};
+
+/// Returns the default exipry period for cached responses
+const fn expiry_default() -> Duration {
+    Duration::from_secs(60 * 60 * 12)
+}
+
+/// Returns the default expiry period for the starting lock
+const fn expiry_starting() -> Duration {
+    Duration::from_secs(20)
+}
+
+/// The data structure containing all necessary components of a response ready to be (de)serialized
+/// from/into the cache
+#[derive(Deserialize, Serialize)]
+enum SerializedResponse {
+    Start,
+    Finished {
+        code: u16,
+        headers: Option<HashMap<String, Vec<u8>>>,
+        body: Option<serde_json::Value>,
```

Comment:
You can't use a json here, it has to be `bytes`, otherwise you may actually return a slightly different response. It's not the end of the world if bytes are much harder, but I believe it's probably not harder?